pkg/karmadactl/interpret/edit.go

```diff
@@ -0,0 +1,511 @@
+		return fmt.Errorf("resource %s/%s is not from file", originalInfo.Namespace, originalInfo.Name)
+	case source == "STDIN" || strings.HasPrefix(source, "http"):
+		w = os.Stdout
+	default:
+		f, err := os.OpenFile(originalInfo.Source, os.O_RDWR|os.O_TRUNC, 0)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		w = f
+
+		_, _, isJSON := yaml.GuessJSONStream(f, 4096)
+		if isJSON {
+			writer = &printers.JSONPrinter{}
+		}
+	}
+
+	err := writer.PrintObj(editedObj, w)
+	if err != nil {
+		fmt.Fprint(o.ErrOut, results.addError(err, originalInfo))
+	}
+
+	printer, err := o.ToPrinter("edited")
+	if err != nil {
+		return err
+	}
+	return printer.PrintObj(originalInfo.Object, o.Out)
+}
+
+func (o *Options) saveToServer(originalInfo *resource.Info, editedObj runtime.Object, results *editResults) error {
+	originalJS, err := json.Marshal(originalInfo.Object)
+	if err != nil {
+		return err
+	}
+
+	editedJS, err := json.Marshal(editedObj)
+	if err != nil {
+		return err
+	}
+
+	preconditions := []mergepatch.PreconditionFunc{
+		mergepatch.RequireKeyUnchanged("apiVersion"),
+		mergepatch.RequireKeyUnchanged("kind"),
+		mergepatch.RequireMetadataKeyUnchanged("name"),
+		mergepatch.RequireKeyUnchanged("managedFields"),
+	}
+
+	patchType := types.MergePatchType
+	patch, err := jsonpatch.CreateMergePatch(originalJS, editedJS)
+	if err != nil {
+		klog.V(4).Infof("Unable to calculate diff, no merge is possible: %v", err)
+		return err
+	}
+	var patchMap map[string]interface{}
+	err = json.Unmarshal(patch, &patchMap)
+	if err != nil {
+		klog.V(4).Infof("Unable to calculate diff, no merge is possible: %v", err)
+		return err
+	}
+	for _, precondition := range preconditions {
+		if !precondition(patchMap) {
+			klog.V(4).Infof("Unable to calculate diff, no merge is possible: %v", err)
+			return fmt.Errorf("%s", "At least one of apiVersion, kind and name was changed")
+		}
+	}
+
+	if o.OutputPatch {
+		fmt.Fprintf(o.Out, "Patch: %s\n", string(patch))
+	}
+
+	patched, err := resource.NewHelper(originalInfo.Client, originalInfo.Mapping).
+		WithFieldManager(o.FieldManager).
+		WithFieldValidation(o.ValidationDirective).
+		WithSubresource(o.Subresource).
+		Patch(originalInfo.Namespace, originalInfo.Name, patchType, patch, nil)
+	if err != nil {
+		fmt.Fprintln(o.ErrOut, results.addError(err, originalInfo))
+		return nil
+	}
+	err = originalInfo.Refresh(patched, true)
+	if err != nil {
+		return err
+	}
+	printer, err := o.ToPrinter("edited")
+	if err != nil {
+		return err
+	}
+	return printer.PrintObj(originalInfo.Object, o.Out)
+}
+
+const (
+	luaCommentPrefix    = "--"
+	luaAnnotationPrefix = "---@"
+	luaAnnotationTarget = luaAnnotationPrefix + "TARGET:"
+	luaAnnotationRule   = luaAnnotationPrefix + "RULE:"
+)
+
+func printCustomization(w io.Writer, c *configv1alpha1.ResourceInterpreterCustomization, rules interpreter.Rules, showDoc bool) {
+	fmt.Fprintf(w, "%s %s %s", luaAnnotationTarget, c.Spec.Target.APIVersion, c.Spec.Target.Kind)
```

Comment:
Would it be better to add a blank line after this?

Reply:
Added